### PR TITLE
[Feature] [PO CSV] Apis to upload validate and operate on csv to crea…

### PR DESCRIPTION
**Pull request by virenchugh25**
…te PO

### CODE REVIEW

File: spec/controllers/api/public/v1/purchase_orders_controller_spec.rb [hunk 1]
```diff
@@ -365,4 +368,171 @@
...
+        File.delete("/tmp/file.csv")
+        expect(response).to have_http_status(:bad_request)
+
+
+        CSV.open("/tmp/file.csv", "wb") do |csv|
+          csv << csv_headers
+          csv << [material_request.id, 'ABC', 100, 'ABC', 1, 1, 1, 1, 1, 1, supplier.id]
+        end
+        post :upload, params: {
+          file: Rack::Test::UploadedFile.new("/tmp/file.csv", "text/csve")
```
Comment (nipunmanocha1mg) [inline spec/controllers/api/public/v1/purchase_orders_controller_spec.rb]: csve?

File: app/models/purchase_order.rb [hunk 2]
```diff
@@ -24,5 +31,6 @@ class PurchaseOrder < ApplicationRecord
   def init
     self.status ||= :draft
```
Comment (kamal2793) [inline app/models/purchase_order.rb]: self.status ||= :created

File: app/services/upload_validations/purchase_order_validation.rb [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    end
+
+    private
+
+    attr_accessor :file_headers, :file
+
+    #
+    # Validates if sku is present
+    #
+    def validate_skus(sku_ids)
```
Comment (kamal2793) [inline app/services/upload_validations/purchase_order_validation.rb]: move this method to base_validator.

File: app/services/upload_validations/purchase_order_validation.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      skus = Sku.where(id: sku_ids)
+      missing_sku_ids = sku_ids - skus.ids
+      return missing_sku_ids
+    end
+
+    #
+    # Validates if suppliers are active
+    #
+    def validate_suppliers(supplier_ids)
```
Comment (kamal2793) [inline app/services/upload_validations/purchase_order_validation.rb]: move this method to base_validator.

File: app/controllers/api/public/v1/purchase_orders_controller.rb [hunk 5]
```diff
@@ -97,5 +111,15 @@ def valid_update?
...
     end
+
+    def valid_upload?
+      param! :file, ActionDispatch::Http::UploadedFile, required: true, blank: false
+      raise BadRequest.new("Invalid file type") if params[:file].content_type != 'text/csv'
+    end
+
+    def valid_force_upload?
+      param! :file, ActionDispatch::Http::UploadedFile, required: true, blank: false
+      raise BadRequest.new("Invalid file type") if params[:file].content_type != 'text/csv'
```
Comment (anubh1mg) [inline app/controllers/api/public/v1/purchase_orders_controller.rb]: valid_upload and valid_force_upload are having same function description

File: app/models/material_request.rb [hunk 6]
```diff
@@ -6,7 +6,7 @@ class MaterialRequest < ApplicationRecord
 
   enum status: {
     created: 'created',
-    pending: 'pending',
+    downloaded: 'downloaded',
     ordered: 'ordered',
     partially_ordered: 'partially_ordered',
```
Comment (anubh1mg) [inline app/models/material_request.rb]: we can remove partially_ordered.

File: app/models/state_transitions/material_request.rb [hunk 7]
```diff
@@ -7,7 +7,7 @@ module MaterialRequest
 
       aasm column: 'status' do
         state :created, initial: true
-        state :pending
+        state :downloaded
         state :ordered
         state :partially_ordered
```
Comment (anubh1mg) [inline app/models/state_transitions/material_request.rb]: remove partially ordered

File: app/services/procurement_module/purchase_order_uploader.rb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      purchase_order_items = []
+
+      CSV.foreach(file.path, headers: true) do |row|
+        supplier_id = row['Supplier Id'].to_i
+        code = row['Code'].to_i
+        sku_id = row['Item Code'].to_i
+        quantity = row['Shortage'].to_i
+
+        if raise_error.blank? && (invalid_values[:missing_supplier_ids].include?(supplier_id) ||
+        invalid_values[:missing_sku_ids].include?(sku_id) || invalid_values[:missing_material_request_ids].include?(code))
```
Comment (anubh1mg) [inline app/services/procurement_module/purchase_order_uploader.rb]: make it in two lines

**Pull request merged**